### PR TITLE
MAID 1361 Added unit tests for id, public_id, data, direct_message, error

### PR DESCRIPTION
**Pull request by mabeebam**
[Review on Reviewable](https://reviewable.io/reviews/maidsafe/routing/680)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @brian-js (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

File: src/id.rs [hunk 1]
```diff
@@ -83,53 +83,54 @@ impl Id {
...
-//
-//         assert!(!relocated.assign_relocated_name(relocated_name.clone()));
-//         assert!(!relocated.assign_relocated_name(Random::generate_random()));
-//         assert!(!relocated.assign_relocated_name(original_name.clone()));
-//
-//
-//         assert!(relocated.is_relocated());
-//         assert_eq!(relocated.name(), relocated_name);
-//         assert!(before.name()!= relocated.name());
-//         assert_eq!(before.signing_public_key(), relocated.signing_public_key());
```
Comment (benjaminbollen) [inline src/id.rs]: would it not be worth preserving these checks as well?

**Comment by benjaminbollen** [comment]
this is a good start!  Can you think of new tests for Id?
can you also chage it to the new code-style: in particular eg avoid `use sodiumoxide;` instead write the full namespace everywhere ```::sodiumoxide::crypto::sign::gen_keypair()`.  We are unfortunately not yet publishing an "official MaidSafe" style guide, but the current gist is here: https://gist.github.com/ustulation/086a62d4f64d02542afa

File: src/id.rs (hunk 1 above)
Comment (mabeebam) [inline src/id.rs]: old 113: is tested in new 110 (dedicated test for fn is_relocated())
old 118: is tested in new 130
old 120: is tested in new 133
old 121: is tested in new 134
old 125: is tested in new 114
old 126: was redundant to old 101 (or should be at least in own unit test)
old 127: correct, should be tested as well (change actually happened)
old 128-132: correct, in order to assure that method does not change value of other properties

Thanks for your feedback!

**Comment by mabeebam** [comment]
Ok, I will add old tests from 126-132 and adapt to style guides.
I could also add unit tests that assure that all getter-functions preserve the value of all properties and constructor does what it should do. You wanna have it refined to such a level?

**Comment by benjaminbollen** [comment]
checking that the values of the keys are not mutated is probably as far as I would go; you can move on to the other items of this task

**Comment by mabeebam** [comment]
Units test for id are done from my point of view. Feedback always welcome!
Why AppVeyor fails is unclear for me.

**Comment by mabeebam** [comment]
Unit test for id.rs are done from my point of view, feedback always welcome! 
Why AppVeyor fails is unclear to me.

**Comment by brian-js** [comment]
Hi @mabeebam, when your happy with this, if you remove WIP from the title I'll take a closer look.

**Comment by mabeebam** [comment]
Hi @brian-js, sure I will remove WIP when I am finished! Looking forward to your feedback!

**Comment by mabeebam** [comment]
Looking forward to feedback on the unit tests. Cheers!

**Comment by brian-js** [comment]
@mabeebam Nice work on this task. The Rust guidelines for line limits is set at 100 characters, can you edit the code to take that into account please, then the PR will be good to merge. Also, add a comment with the BTC address we should send your payment to.

**Comment by mabeebam** [comment]
@brian-js Great, thanks for your feedback. I will limit lines to 100 chars. Let you know when I have done that.

**Comment by mabeebam** [comment]
@brian-js Refactoring is done!

**Comment by brian-js** [comment]
Thanks @mabeebam. If you let us know the BTC address to send the payment to we'll get it to you asap.

**Comment by dirvine** [comment]
Ohhps @mabeebam  A slight issue with accumulator I think. Will check it out in the morning.

**Comment by mabeebam** [comment]
@dirvine let me know if I have to fix anything!

**Comment by brian-js** [comment]
@mabeebam Nothing to fix from your end. The accumulator got an update and published so all should be good.

**Comment by mabeebam** [comment]
BTC address = 17EZ8xNLzZmHKiFtkCnx5xAC4wNjZ6PyTX

**Comment by NickLambert** [comment]
Bounty payment on the way to you @mabeebam. Thanks for all you help.

**Comment by mabeebam** [comment]
@NickLambert Thanks, got it! Hope to help soon again!

**Pull request merged**